Tests/MongoSwiftSyncTests/UnifiedTestRunner/UnifiedOperation.swift

```diff
@@ -6,6 +6,39 @@ import TestsCommon
 protocol UnifiedOperationProtocol: Decodable {
     /// Set of supported arguments for the operation.
     static var knownArguments: Set<String> { get }
+
+    /// Indicates whether this operation returns an array of root documents. This is used when determining how to match
+    /// against the operation's expected result.
+    static var returnsRootDocuments: Bool { get }
```

Comment:
I wonder if it makes more sense to make `UnifiedOperatoinResult` a little more specific and store this on the result itself rather than on the operation.

e.g. 
```swift
enum UnifiedOperationResult {
    case changeStream(...)
    case value(BSON) // this could be broken down further if we want also
    case rootDocumentArray([BSONDocument])
}
```
I'm finding the various booleans passed around in the `matches` function to be a little confusing, and this data-oriented approach makes a little more sense to me. The cost here is introducing more specificity at the operation execution level, which I don't think should be a big deal. What do you think?

Reply:
I think the new `MatchableResult` looks great! I left a few other comments with minor suggestions about it, but otherwise I think its a nice improvent.

In regards to the new enum, I think it would be mostly useful as a way to distinguish `specialOperator` from the rest of the cases, but I don't necessarily see much value in differentiating from among the various BSON cases. I think the `BSON` enum is sufficient for that.

e.g. something like

```swift
enum ExpectedResult {
    case specialOperator(SpecialOperator)
    case bson(BSON)
}

enum SpecialOperator {
    case exists(key: String)
    case type(type: String)
    case matchesEntity(entityId: String)
    ...
}
```
I think these type(s) could be useful for making the `matchesSpecial` code a little cleaner via `Codable`. I don't think this would be _necssary_, but I do think it would be a minor improvement, especially if the special matches get more complicated / involve more data down the road. What do you think?

Reply:
sorry, couldn't reply to this with my previous batch:

I tried out the `ExpectedResult` approach, however I found it rather clunky because I had to explicitly initialize `ExpectedResult`s in the places where we call `matchesInner` (now renamed just `matches`). It also introduced another level of switching in `matches` since we have to first switch on the `ExpectedResult` and then the nested `BSON`.

what I ended up doing is still adding a `SpecialOperator` type, which we just directly initialize and use when we see a special operator doc. I think it makes the special matching code a bit cleaner without making the rest of the code more complicated; let me know what you think.